src/app/qgsmaptoolselectutils.cpp

```diff
@@ -335,3 +344,383 @@ QgsFeatureIds QgsMapToolSelectUtils::getMatchingFeatures( QgsMapCanvas *canvas,
+{
+  if ( featureCount == 0 || featureCount == 1 )
+  {
+    switch ( mBehavior )
+    {
+      case QgsVectorLayer::SetSelection:
+        return tr( "Select" );
+        break;
+      case QgsVectorLayer::AddToSelection:
+        return tr( "Add to Selection" );
+        break;
+      case QgsVectorLayer::IntersectSelection:
+        return tr( "Intersect with Selection" );
+        break;
+      case QgsVectorLayer::RemoveFromSelection:
+        return tr( "Remove from Selection" );
+        break;
+    }
+  }
+
+  QString featureCountText;
+  if ( featureCount < 0 )
+    featureCountText = QStringLiteral( "Searching..." );
+  else
+    featureCountText = QString::number( featureCount );
+
+  switch ( mBehavior )
+  {
+    case QgsVectorLayer::SetSelection:
+      return tr( "Select All (%1)" ).arg( featureCountText );
+      break;
+    case QgsVectorLayer::AddToSelection:
+      return tr( "Add All to Selection (%1)" ).arg( featureCountText );
+      break;
+    case QgsVectorLayer::IntersectSelection:
+      return tr( "Intersect All with Selection (%1)" ).arg( featureCountText );
+      break;
+    case QgsVectorLayer::RemoveFromSelection:
+      return tr( "Remove All from Selection (%1)" ).arg( featureCountText );
+      break;
+  }
+
+  return QString();
+}
+
+QString QgsMapToolSelectUtils::QgsMapToolSelectMenuActions::textForChooseOneMenu() const
+{
+  switch ( mBehavior )
+  {
+    case QgsVectorLayer::SetSelection:
+      return tr( "Select One" );
+      break;
+    case QgsVectorLayer::AddToSelection:
+      return tr( "Add One to Selection" );
+      break;
+    case QgsVectorLayer::IntersectSelection:
+      return tr( "Intersect One with Selection" );
+      break;
+    case QgsVectorLayer::RemoveFromSelection:
+      return tr( "Remove One from Selection" );
+      break;
+  }
+
+  return QString();
+}
+
+
+void QgsMapToolSelectUtils::QgsMapToolSelectMenuActions::populateChooseOneMenu( const QgsFeatureIds &ids )
+{
+  QgsFeatureIds displayedFeatureIds;
+
+  QgsFeatureIds::ConstIterator it = ids.constBegin();
+  while ( displayedFeatureIds.count() <= 20 && it != ids.constEnd() ) //for now hardcoded, but maybe define a settings for this
+    displayedFeatureIds.insert( *( it++ ) );
+
+  for ( QgsFeatureId id : qgis::as_const( displayedFeatureIds ) )
+  {
+    QAction *featureAction = new QAction( tr( "Feature %1" ).arg( id ), this ) ;
+    featureAction->setData( id );
+    connect( featureAction, &QAction::triggered, this, &QgsMapToolSelectMenuActions::chooseOneCandidateFeature );
+    connect( featureAction, &QAction::hovered, this, &QgsMapToolSelectMenuActions::highlightFeatures );
+    mMenuChooseOne->addAction( featureAction );
+  }
+
+  mMenuChooseOne->setEnabled( ids.count() != 0 );
+}
+
+void QgsMapToolSelectUtils::QgsMapToolSelectMenuActions::chooseOneCandidateFeature()
+{
+  if ( !mVectorLayer )
+    return;
+  QAction *senderAction = qobject_cast<QAction *>( sender() );
```

Comment:
it's a cleaner approach to use a lambda connection and avoid the sender/data dance

Reply:
done